doc/sphinx/proofs/writing-proofs/equality.rst

```diff
@@ -31,6 +31,44 @@ There are multiple notions of :gdef:`equality` in Coq:
   other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
   two terms are definitionally equal.
 
+Equality
+--------
```

Comment:
For this initial section, we obviously need a better title. I was thinking something like "Basic tactics for Leibniz and setoid equality".

And like my title suggestion indicates, the current descriptions of these tactics is too limited. It should be clarified that `reflexivity`, `symmetry` and `transitivity` also work for any equivalence relation (registered with the `Equivalence` typeclass) or for any reflexive, symmetric or transitive relation (registered with the `Reflexive`, `Symmetric` and `Transitive` typeclasses).

Reply:
> ... also work for any equivalence relation

In other words, using setoid equality, right?

The typeclasses chapter doesn't explain how to register typeclasses.  Indeed, it lacks an introduction, instead pointing the user to a paper "Matthieu Sozeau and Nicolas Oury. First-Class Type Classes. In TPHOLs'08. 2008." (is it worth trying to include links to papers or is it understood that you look on the author's website?)  And surely the concept can be explained more simply and in far less than 16 pages.

At the least, I think an example or two of registering typeclasses is needed in that chapter.  It would be better if we provide a succinct summary of typeclasses.  Would you care to identify the main points for a summary?

The syntax in the first lines of the typeclasses chapter should use our Sphinx notation.  `Class` and `Instance` should be much earlier in the chapter.  I don't like dumping all the commands at the end of the chapter.

Also, I don't recall seeing a section relating typeclass resolution to other standard processing steps (i.e. no good overview of the steps in order).

But OTOH I don't want to spend too much time on the typeclasses chapter in this PR.

Reply:
> "Basic tactics for Leibniz and setoid equality"

How about "Reflexivity, symmetry and transitivity tactics"?  Specific and I think quite clear.  "Basic" kinda forces you to examine the text to understand what is covered.

Reply:
Are there any other predefined typeclasses that tactics rely on?  The typeclasses chapter should mention all of them.

Reply:
Your title doesn't include `f_equal`. And adding one more to the list becomes really heavy IMHO.

>In other words, setoid but not necessarily equality, right?

Setoid equalities are equivalence relations...

Yes, the typeclass chapter is terrible and needs to be improved. But not in this PR please.

Reply:
> Your title doesn't include `f_equal`.

It's not a perfect title but I think it's quite good.  `f_equal` is not exactly a household name.  It's too easy to overuse vague terms such as "basic".  But I can apply your suggestion.  Aside from that, perhaps this is ready to squash and merge?